src/datasets/arrow_dataset.py

```diff
@@ -1179,63 +1181,73 @@ def __exit__(self, exc_type, exc_val, exc_tb):
 
         Args:
-            dataset_path (:obj:`str`): Path (e.g. `dataset/train`) or remote URI (e.g. `s3://my-bucket/dataset/train`)
+            dataset_path (``PathLike``): Path (e.g. `path/to/dataset/train`) or remote URI (e.g. `s3://my-bucket/dataset/train`)
                 of the dataset directory where the dataset will be saved to.
-            fs (:class:`~filesystems.S3FileSystem`, ``fsspec.spec.AbstractFileSystem``, optional, defaults ``None``):
-                Instance of the remote filesystem used to download the files from.
+            num_shards (:obj:`int`, optional): Number of shards to write.
+                Default to the same value as `num_proc` if specified.
+
+                <Added version="2.8.0"/>
+            num_proc (:obj:`int`, optional, default `None`): Number of processes when downloading and generating the dataset locally.
+                Multiprocessing is disabled by default.
+
+                <Added version="2.8.0"/>
+            storage_options (:obj:`dict`, *optional*): Key/value pairs to be passed on to the file-system backend, if any.
+
+                <Added version="2.8.0"/>
 
         Example:
 
         ```py
         >>> saved_ds = ds.save_to_disk("path/to/dataset/directory")
         ```
         """
+        num_proc: int = num_proc if num_proc is not None else 1
+        num_shards: int = num_shards if num_shards is not None else num_proc
+        if fs != "deprecated":
+            warnings.warn(
+                "'fs' was is deprecated in favor of 'storage_options' in version 2.8.0 and will be removed in 3.0.0.\n"
+                "You can remove this warning by passing 'storage_options=fs.storage_options' instead.",
+                FutureWarning,
+            )
+            storage_options = fs.storage_options
+
+        fs_token_paths = fsspec.get_fs_token_paths(dataset_path, storage_options=storage_options)
+        fs: fsspec.AbstractFileSystem = fs_token_paths[0]
+        is_local = not is_remote_filesystem(fs)
+        path_join = os.path.join if is_local else posixpath.join
+
         if self.list_indexes():
             raise ValueError("please remove all the indexes using `dataset.drop_index` before saving a dataset")
 
-        dataset = self.flatten_indices() if self._indices is not None else self
+        dataset = self.flatten_indices(num_proc=num_proc) if self._indices is not None else self
 
-        if is_remote_filesystem(fs):
-            dataset_path = extract_path_from_uri(dataset_path)
-        else:
-            fs = fsspec.filesystem("file")
-            cache_files_paths = [Path(cache_filename["filename"]) for cache_filename in self.cache_files]
+        if is_local:
+            Path(dataset_path).resolve().mkdir(parents=True, exist_ok=True)
+            parent_cache_files_paths = set(
+                Path(cache_filename["filename"]).resolve().parent for cache_filename in self.cache_files
+            )
```

Comment:
I'm okay with checking the parent dir instead of the actual cache filename, but why is this change needed?

Reply:
It's just simpler to implelement this verification this way than iterate over the file names x)
And it also prevents the case where you'd have the dataset saved twice in the same folder, but with different number of shards e.g.
```
data-00000-of-00008.parquet
data-00001-of-00008.parquet
...
data-00007-of-00008.parquet
```
and

```
data-00000-of-00004.parquet
data-00001-of-00004.parquet
...
data-00003-of-00004.parquet
```